## Download and prepare data

----------------------------------

**Open files and examine contents**

In [1]:
import math
import pandas as pd
import numpy as np
from scipy import stats as st
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
df_0 = pd.read_csv('/datasets/geo_data_0.csv')
df_1 = pd.read_csv('/datasets/geo_data_1.csv')
df_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [3]:
df_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
df_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [5]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [6]:
df_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [7]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [8]:
df_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [9]:
print(df_0.duplicated().sum())
print(df_1.duplicated().sum())
print(df_2.duplicated().sum())

0
0
0


I first imported the datasets for each region and did some inital exploration of the contents. I also made sure there were no duplicate observations.

**Preprocessing of data, training and test sets**

In [10]:
target_df0 = df_0['product']
features_df0 = df_0.drop(['product','id'], axis=1)

In [11]:
target_df1 = df_1['product']
features_df1 = df_1.drop(['product','id'], axis=1)

In [12]:
target_df2 = df_2['product']
features_df2 = df_2.drop(['product','id'], axis=1)

In [13]:
features_train_df0, features_valid_df0, target_train_df0, target_valid_df0 = train_test_split(
    features_df0, target_df0, test_size=0.25, random_state=12345) 

print(features_train_df0.shape, features_valid_df0.shape)
print(target_train_df0.shape, target_valid_df0.shape)

(75000, 3) (25000, 3)
(75000,) (25000,)


In [14]:
scaler = StandardScaler()

scaler.fit(features_train_df0)
scaler.transform(features_train_df0)

array([[-0.5448279 ,  1.39026372, -0.09495893],
       [ 1.4559119 , -0.48042154,  1.20956708],
       [ 0.26045969,  0.82506858, -0.2048645 ],
       ...,
       [ 0.41894874, -1.29678805, -0.19640667],
       [ 0.40007671, -1.46687406, -0.44531736],
       [ 1.746246  ,  0.02741521,  2.76684795]])

In [15]:
features_train_df1, features_valid_df1, target_train_df1, target_valid_df1 = train_test_split(
    features_df1, target_df1, test_size=0.25, random_state=12345) 

print(features_train_df1.shape, features_valid_df1.shape)
print(target_train_df1.shape, target_valid_df1.shape)

(75000, 3) (25000, 3)
(75000,) (25000,)


In [16]:
scaler.fit(features_train_df1)
scaler.transform(features_train_df1)

array([[-0.85085526,  0.62442838,  0.29694289],
       [ 1.97193524,  1.83227487,  0.29433274],
       [ 1.07930536,  0.17012731, -0.29641817],
       ...,
       [ 1.04707022, -0.64999232,  1.47336769],
       [-0.11478048, -1.19069924,  0.29915578],
       [-0.64614644,  0.09907453,  0.29561113]])

In [17]:
features_train_df2, features_valid_df2, target_train_df2, target_valid_df2 = train_test_split(
    features_df2, target_df2, test_size=0.25, random_state=12345) 


print(features_train_df2.shape, features_valid_df2.shape)
print(target_train_df2.shape, target_valid_df2.shape)

(75000, 3) (25000, 3)
(75000,) (25000,)


In [18]:
scaler.fit(features_train_df2)
scaler.transform(features_train_df2)

array([[-0.52615957,  0.77632883, -0.40079292],
       [-0.88962499, -0.4040698 , -1.22293566],
       [-1.1339838 ,  0.20857647,  0.2967648 ],
       ...,
       [ 0.36856367,  0.79722508,  0.66557457],
       [-2.44068989,  0.11378427,  0.440907  ],
       [-1.73246811,  0.39357329, -1.4244654 ]])

To prepare my data, I selected the relevant features, elimnating the `id` column, and established my target value. I also standardized the features I would be using, and split each dataframe into a training a validation set with a 3:1 ratio. 

## Train and test model for each reigon

-------------------------------

In [19]:
model = LinearRegression()

In [59]:
def train_models(features_train, features_valid, target_train, target_valid):
    model.fit(features_train, target_train)
    
    predictions = model.predict(features_valid)
    correct = target_valid
    
    return predictions, correct

**Region 1:**

In [58]:
r1_predictions, r1_correct = train_models(
                features_train_df0, features_valid_df0, target_train_df0, target_valid_df0)

mse = mean_squared_error(r1_correct, r1_predictions)

print('Average volume of predicted reserves (in thousands of barrels): {:.4f}'.format(
                                                                        r1_predictions.mean()))
print('Model RMSE: {:.2f}'.format(mse ** 0.5))

Average volume of predicted reserves (in thousands of barrels): 92.5926
Model RMSE: 37.58


In [22]:
print(target_df0.describe())

count    100000.000000
mean         92.500000
std          44.288691
min           0.000000
25%          56.497507
50%          91.849972
75%         128.564089
max         185.364347
Name: product, dtype: float64


**Region 2:**

In [23]:
r2_predictions, r2_correct = train_models(
                features_train_df1, features_valid_df1, target_train_df1, target_valid_df1)

mse = mean_squared_error(r2_correct, r2_predictions)

print('Average volume of predicted reserves (in thousands of barrels): {:.4f}'.format(
                                                                        r2_predictions.mean()))
print('Model RMSE: {:.2f}'.format(mse ** 0.5))

Average volume of predicted reserves (in thousands of barrels): 68.7285
Model RMSE: 0.89


In [24]:
print(target_df1.describe())

count    100000.000000
mean         68.825000
std          45.944423
min           0.000000
25%          26.953261
50%          57.085625
75%         107.813044
max         137.945408
Name: product, dtype: float64


**Region 3:**

In [25]:
r3_predictions, r3_correct = train_models(
                features_train_df2, features_valid_df2, target_train_df2, target_valid_df2)

mse = mean_squared_error(r3_correct, r3_predictions)

print('Average volume of predicted reserves (in thousands of barrels): {:.4f}'.format(
                                                                        r3_predictions.mean()))
print('Model RMSE: {:.2f}'.format(mse ** 0.5))

Average volume of predicted reserves (in thousands of barrels): 94.9650
Model RMSE: 40.03


In [26]:
print(target_df2.describe())

count    100000.000000
mean         95.000000
std          44.749921
min           0.000000
25%          59.450441
50%          94.925613
75%         130.595027
max         190.029838
Name: product, dtype: float64


**Analysis:**

To train and test my model, I wrote a function that I could use with each region. I then found both the average volume of reserves my model predicted, and the RMSE. Since I found the actual average of each region's target value, I can see that on first glance they look fairly close to the predictions. 

**Region 1** actual mean is ~92.5001, while the mean for my model's predictions is ~95.5926. 

**Region 2** actual mean is ~68.8250  and predicted mean is ~68.7285. 

**Region 3** actual mean is ~95.0004 and predicted mean is ~94.9650. 

I also found the RMSE, to evaluate how accurately my model was predicting well reserves. Region 2  appears to be the easiest to predict accurately, with an average error equating to <1 unit. Both region 1 and region 3 RMSE is much higher, 37.58 units and 40.03 units, respectively. This value represents somewhere between 170k and 180k USD of revenue, and could mean that projections about well reserves and future profits in these two regions are less reliable. The mathematical descriptions of region 1 and 3 reveal a possible cause of this uncertainty, as there is more dispersion across their reserves compared to region 2. 

## Prepare for profit calculations

----------------------------------------

In [27]:
BUDGET_PER_WELL = 100000000 / 200 # USD

REVENUE_PER_UNIT = 4500 # USD

required_reserves = BUDGET_PER_WELL / REVENUE_PER_UNIT

print('Minimum reserves required per well to develop without losses, in thousands of barrels:')
print('{:.2f}'.format(required_reserves))

Minimum reserves required per well to develop without losses, in thousands of barrels:
111.11


In [28]:
print('Average volume of reserves, per well (in thousands of barrels)')
print('Region 1: {:.2f}'.format(df_0['product'].mean()))
print('Region 2: {:.2f}'.format(df_1['product'].mean()))
print('Region 3: {:.2f}'.format(df_2['product'].mean()))

Average volume of reserves, per well (in thousands of barrels)
Region 1: 92.50
Region 2: 68.83
Region 3: 95.00


**Analysis:** 

To find how much a well needs to have in its reserves to be profitable or at least break even, I calculated the budget that could be allocated to each well, and divided it by the price per unit. I found that a well would need at least ~111.11 thousand barrels. Comparing this against the average volume per well in each region, it seems possible that regions 1 and 3 are more likely to have wells with sufficient reserves to avoid losses. The mathematical descriptions found earlier backs up this conclusion, as they show that region 2's data points representing each quartile are consistently lower than region 1 and 3. For example, region 2's median is about the same as region 1 and 3's first quartile. So, while it may be easier to accurately predict the reserves in region 2, the overall volume of reserves in the region are significantly lower. 

## Calculate profit from selected oil wells and model predictions

-------------------------

**Top 200 highest predicted reserves, by region**

In [29]:
top_200_r1 = pd.Series(r1_predictions).sort_values(axis=0, ascending=False)[:200]
top_200_r2 = pd.Series(r2_predictions).sort_values(axis=0, ascending=False)[:200]
top_200_r3 = pd.Series(r3_predictions).sort_values(axis=0, ascending=False)[:200]

In [30]:
print(top_200_r1.mean())
print(top_200_r2.mean())
print(top_200_r3.mean())

155.51165419405697
138.73013391081713
148.01949329159174


**Function for calculating total profits from development of 200 wells**

In [44]:
def total_profits(targets, predictions):
    top = predictions.sort_values(axis=0, ascending=False)[:200]
    top_200_target = targets[top.index]
    
    result = top_200_target.sum() * 4500
    return result - 100000000

In [52]:
print('Total profits from wells selected by top 200 predictions:')
print('Region 1: ${:.2f}'.format(
    total_profits(target_df0, pd.Series(model.predict(features_df0)))))
print('Region 2: ${:.2f}'.format(
    total_profits(target_df1, pd.Series(model.predict(features_df1)))))
print('Region 3: ${:.2f}'.format(
    total_profits(target_df2, pd.Series(model.predict(features_df2)))))

Total profits from wells selected by top 200 predictions:
Region 1: $27686098.34
Region 2: $21389816.37
Region 3: $25714106.32


**Analysis**

Since the 200 wells selected to develop are presumably the ones with the most oil reserves, I found the top 200 predictions (by volume) for each region. I also calculated their respective means to see what the target volume of reserves is when selecting wells to develop, about 140 thousand barrels. The function `total_profits` takes an input of an arbitrary number of wells, finds the top 200 predictions by volume, finds the actual wells that these predictions correspond to, then calculates what the total profit would be from developing them. After this analysis, my initial recommendation for the region to develop is region 1 with the highest profit from its top 200 predicted wells of all three, and the largest average predicted well reserves. However, regions 1 and 3 were also the two with the least accurate predictions, and region 2 is not far behind in total profits for its top 200 predicted wells, so further investigation could indicate otherwise.

## Analysis of risks and profit for each region

-------------------------------------

**Find distribution of profits with bootstrapped samples from each region**

In [32]:
state = np.random.RandomState(12345)

In [45]:
# bootstrap function to find distribution of profit

def bootstrap(df):
    region_samples = []
    for i in range(1000):
        subsample = df.sample(n=500, replace=True, random_state=state)
        targets = subsample['product']
        predictions = model.predict(subsample[['f0','f1','f2']])
        region_samples.append(total_profits(targets, pd.Series(predictions, index=targets.index)))
    return region_samples

In [46]:
region_1_profits = pd.Series(bootstrap(df_0))
region_2_profits = pd.Series(bootstrap(df_1))
region_3_profits = pd.Series(bootstrap(df_2))


r1_prf_mean = region_1_profits.mean()
r2_prf_mean = region_2_profits.mean()
r3_prf_mean = region_3_profits.mean()

r1_conf_int = (region_1_profits.quantile(q=0.025), region_1_profits.quantile(q=0.975))
r2_conf_int = (region_2_profits.quantile(q=0.025), region_2_profits.quantile(q=0.975))
r3_conf_int = (region_3_profits.quantile(q=0.025), region_3_profits.quantile(q=0.975))

In [47]:
print('Average total profit of samples and cofidence interval')
print('Region 1:')
print(r1_prf_mean, r1_conf_int)
print('Region 2:')
print(r2_prf_mean, r2_conf_int)
print('Region 3:')
print(r3_prf_mean, r3_conf_int)

Average total profit of samples and cofidence interval
Region 1:
3363341.1490838462 (-2288043.605258777, 8600450.30659621)
Region 2:
4648937.162503721 (529858.3298823935, 9108568.167597998)
Region 3:
4220104.2371187955 (-1069840.1964871965, 9660332.02093078)


**Evaluate risk of losses for each region**

In [60]:
print('Risk of losses for Region 1: {:%}'.format(np.mean(region_1_profits < 0)))
print('Risk of losses for Region 2: {:%}'.format(np.mean(region_2_profits < 0)))
print('Risk of losses for Region 3: {:%}'.format(np.mean(region_3_profits < 0)))

Risk of losses for Region 1: 10.100000%
Risk of losses for Region 2: 1.700000%
Risk of losses for Region 3: 5.300000%


**Analysis**

To further analyze both the risks and potential profits developing in each region could result in, I wrote a function that used bootstrapping to model 1000 different scenarios for each region. Each sample had 500 wells, selected randomly from the target values of the original datasets, which I passed to the `total_profits` function. This resulted in a list of the profits that each scenario would net, which I could then use to find both an average and a confidence interval for each region. I also found the risk of losses (negative profits) for all three. While region 2 appeared to be potentially less profitable initially, with a lower average volume of reserves and lower predicted profits for the models top 200 predictions, it seems that a model can make much more accurate predictions for this region. The higher profits for region 1 and 3 were a better reflection of how overstated the model's predictions were, rather than actual profitability. The bootstrapping I did to find profit distribution and risk of losses show that region 2 had the highest average profit from 1k samples, and the lowest risk of losses by far. Additionally, region 1 and 3 had negative profits of about -1m and -2.3m respectively within the 95% confidence interval. After this analysis, my final suggestion is to develop region 2. 